workflow-runtime/src/commonMain/kotlin/com/squareup/workflow1/internal/WorkStealingDispatcher.kt

```diff
@@ -0,0 +1,306 @@
+    // This will either synchronously run block or dispatch it, depending on what resuming a
+    // continuation on the delegate dispatcher would do.
+    continuation?.resumeOnDelegateDispatcher()
+  }
+
+  /**
+   * Throws [UnsupportedOperationException]. We can't allow the default implementation to run in
+   * case the delegate dispatcher would throw.
+   */
+  @ExperimentalCoroutinesApi
+  override fun limitedParallelism(parallelism: Int): CoroutineDispatcher {
+    // We could probably support this by forwarding the call to the delegate then wrapping that
+    // dispatcher with a WSD that advances when this one does, but we don't need this for our use
+    // cases and getting all the behavior correct might be hard, so don't bother for now.
+    throw UnsupportedOperationException(
+      "limitedParallelism is not supported for WorkStealingDispatcher"
+    )
+  }
+
+  /**
+   * "Steals" work that was scheduled on this dispatcher but hasn't had a chance to run yet and runs
+   * it, until there is no work left to do. If the work schedules more work, that will also be ran
+   * before the method returns.
+   *
+   * Some care must be taken with this method: if it is called from a coroutine that is dispatched
+   * by this dispatcher, then it may run inside the stack frame of another call and in that case the
+   * inner call may think the dispatcher is idle when the outer call still has more tasks to run.
+   *
+   * It is technically safe to call from multiple threads, even in parallel, although the behavior
+   * is undefined. E.g. One thread might return from this method before the other has finished
+   * running all tasks.
+   */
+  // If we need a strong guarantee for reentrant calls we could use a ThreadLocal so the inner call
+  // could steal work from the outer one.
+  // If we need a strong guarantee for calling from multiple threads we could just run this method
+  // with a separate lock so all threads would just wait on the first one to finish running, but
+  // that could deadlock if any of the dispatched coroutines call this method reentrantly.
+  fun advanceUntilIdle() {
+    var wasDispatchScheduled = false
+    advanceUntilIdle(
+      onStartLocked = {
+        // If no dispatch was scheduled, then set the flag so that any new dispatch calls that
+        // happen while we're draining the queue won't schedule one unnecessarily since we'll just
+        // handle them.
+        // Note that we could "cancel" the dispatch if this is true here, since we're stealing all
+        // its work, but we can't cancel that task so it will just have to noop.
+        wasDispatchScheduled = dispatchScheduled.also {
+          if (!it) dispatchScheduled = true
+        }
+      },
+      onFinishedLocked = {
+        // If we set this flag above, then clear it now so future dispatch calls schedule normally.
+        dispatchScheduled = wasDispatchScheduled
```

Comment:
why would we ever need to set it back to true? haven't we just drained the queue?